aten/src/ATen/core/autocast/register_autocast_ops.cpp

```diff
@@ -0,0 +1,528 @@
+#include <ATen/core/op_registration/op_registration.h>
+#include <ATen/ATen.h>
+#include <ATen/NativeFunctions.h>
+// TODO:  Remove this dependency once callBoxed is possible for all ops
+#include <torch/csrc/jit/operator.h>
+
+#include <iostream>
+#include <exception>
+
+namespace at {
+namespace autocast {
+
+namespace {
+// Imitate Apex and cache some of the casts to streamline parameter reuse.
+// Our heuristic is to cache FP16 casts of FP32 model weights (see cached_cast below).
+//
+// After discussion with @ezyang, the cache uses the following structure:
+// The key is the source tensor's TensorImpl*, a proxy for a Tensor uuid that's unchanged across shallow copies.
+// The value is a tuple of two Tensors:  the source tensor and the casted tensor.  Both are stored as full
+// Tensors to ensure their TensorImpls stay alive if other references to them die.
+//
+// We must keep the source tensor alive because if the source tensor were deallocated, another random Tensor
+// could be allocated whose TensorImpl* happened to have the same value.  This TensorImpl* would then mistakenly
+// hit in cache, which would be a nasty bug (rare, intermittent, unpredictable).
+//
+// Since our heuristic caches casts for model weights only, the source Tensors should always stay alive anyway,
+// because the model stores them explicitly.  So in the common case, storing a live reference to the
+// source tensor in cached_casts is unnecessary but also does no harm (does not increase memory use).
+//
+// When the autocast context manager exits, which should occur at the end of each forward pass, it calls
+// clear_cache to ensure cached Tensors don't leak outside the autocasting region.
+thread_local std::unordered_map<TensorImpl*, std::tuple<Tensor, Tensor>> cached_casts;
+}
+
+void clear_cache() {
+  cached_casts.clear();
+}
+
+enum class CastPolicy : uint8_t {
+  fp16 = 0, // Cast all inputs to fp16 before running the op.
+  fp16_with_tensorlist, // Separate from CastPolicy::fp16 due to lifetime concerns of data
+                        // viewed by TensorLists.  Maybe some C++ guru can help me unify them.
```

Comment:
So what exactly is going on here anyway?

Reply:
This is tied to the vector lifetime issues we're already discussing.  For ops like cat and stack where there's an incoming TensorList that I need to replace with a vector of casted tensors, I decided to create the casted vector as an lvalue on its own line before redispatching, to ensure its data's lifetime lasts the full duration of the redispatch call.  This is divergent code from the simple `CastPolicy::fp16` and `CastPolicy::fp32` cases so it gets its own policy and specialization.

If it's ok to create casted vectors temporaries directly inline in the redispatch call's arguments, I can unify the `*_with_tensorlist` policy with the simple `fp32` and `fp16` cases.  That would be nicer codewise.

Reply:
Rewrote to create vector temporaries inline, deleted `*_with_tensorlist` as independent cases.